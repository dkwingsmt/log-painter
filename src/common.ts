import Color from 'color';

export interface AnalysedLine {
  playerId: string;
  time?: string;
  title?: string;
  content: string[];
}

export interface AnalysedPlayer {
  playerId: string;
  allPlayerIds: string[];
  number?: string;
  name: string;
}

export interface ConfigPlayer {
  playerId: string;
  displayName: string;
  enabled: boolean;
  color: string;
}

export interface Configuration {
  players: Record<string, ConfigPlayer>;
}

export interface DescribedColor {
  value: string;
  name?: string;
  isLight: boolean;
}

export const presetDescribedColors: DescribedColor[] = [
  { value: "black", name: "黑色" },
  { value: "silver", name: "灰色" },
  { value: "red", name: "红色" },
  { value: "green", name: "绿色" },
  { value: "orange", name: "橘色" },
  { value: "purple", name: "紫色" },
  { value: "teal", name: "蓝绿" },
  { value: "fuchsia", name: "桃红" },
  { value: "yellow", name: "黄色" },
  { value: "beige", name: "米色" },
  { value: "brown", name: "棕色" },
  { value: "navy", name: "深蓝" },
  { value: "maroon", name: "紫红" },
  { value: "limegreen", name: "莱姆" },
  { value: "white", name: "白色" },
  { value: "brown", name: "蓝色" },
  { value: "pink", name: "粉红" },
].map(({ value, ...others }) => ({
  ...others,
  value,
  isLight: Color(value).isLight(),
}));