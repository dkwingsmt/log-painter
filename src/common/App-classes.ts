import { makeStyles, createStyles } from '@material-ui/core/styles';

export const useStepperStyles = makeStyles((/*theme: Theme*/) =>
  createStyles({
    Control: {
      display: 'flex',
      marginTop: 20,
    },
    Body: {
    },
    Container: {
      paddingTop: 10,
    },
    ControlButton: {
      width: 100,
      marginLeft: 10,
    }
  }),
);
