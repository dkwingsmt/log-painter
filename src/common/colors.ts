import Color from 'color';

import fromPairs from 'lodash/fromPairs';
import once from 'lodash/once';

export interface DescribedColor {
  value: string;
  name?: string;
  isLight: boolean;
}

interface RawColor {
  value: string;
  name: string;
}

function initializeColors(rawData: RawColor[]): Record<string, DescribedColor> {
  return fromPairs(rawData.map(
    ({ value, ...others }) => [value, {
      ...others,
      value,
      isLight: Color(value).isLight(),
    }],
  ));
}

const rawBbsColors = [
  { value: "black", name: "黑色" },
  { value: "silver", name: "灰色" },
  { value: "red", name: "红色" },
  { value: "green", name: "绿色" },
  { value: "orange", name: "橘色" },
  { value: "purple", name: "紫色" },
  { value: "teal", name: "蓝绿" },
  { value: "fuchsia", name: "桃红" },
  { value: "yellow", name: "黄色" },
  { value: "beige", name: "米色" },
  { value: "brown", name: "棕色" },
  { value: "navy", name: "深蓝" },
  { value: "maroon", name: "紫红" },
  { value: "limegreen", name: "莱姆" },
  { value: "white", name: "白色" },
  { value: "brown", name: "棕色" },
  { value: "blue", name: "蓝色" },
  { value: "pink", name: "粉红" },
];

const rawV2Colors = [
  { value: "black", name: "黑色" },
  { value: "silver", name: "银灰" },
  { value: "#634200", name: "深褐" },
  { value: "#e39700", name: "暗橙" },
  { value: "#d4e300", name: "金色" },
  { value: "#1ee300", name: "草绿" },
  { value: "#26067d", name: "靛色" },
  { value: "#c20cf0", name: "蓝紫" },
  { value: "#fc6d0d", name: "阳橙" },
  { value: "#0dfccc", name: "蓝绿" },
  { value: "#095157", name: "暗岩灰" },
  { value: "#8a0e1e", name: "暗红" },
  { value: "#e31717", name: "鲜红" },
  { value: "#1717e3", name: "蓝色" },
  { value: "#fc1956", name: "樱桃红" },
  { value: "#bd1c72", name: "品红" },
  { value: "#b06635", name: "锗黄" },
  { value: "#315723", name: "暗绿" },
  { value: "#bda855", name: "暗卡其" },
  { value: "#9a55bd", name: "中蓝紫" },
  { value: "#5ebd5e", name: "叶绿" },
  { value: "#65a8c9", name: "灰蓝" },
  { value: "#6b88d6", name: "矢车菊蓝" },
  { value: "#68bdac", name: "中碧蓝" },
  { value: "#575034", name: "咖啡" },
  { value: "#d68c81", name: "玫瑰褐" },
  { value: "#3d4057", name: "普鲁士蓝" },
  { value: "#bd84a2", name: "火鹤红" },
  { value: "#574141", name: "暗灰" },
  { value: "#ebf0d8", name: "米黄" },
  { value: "#eae3fc", name: "薰衣草紫" },
];

export type ColorPalette = 'bbs' | 'v2';

export interface PaletteInfo {
  id: ColorPalette;
  name: string;
  description: string;
  contents: () => Record<string, DescribedColor>,
}

export const colorPalettes: Record<ColorPalette, PaletteInfo> = {
  'v2': {
    id: 'v2',
    name: '色板v2（推荐）',
    description: '“色板v2”包括31个精心挑选的颜色，增强了色彩间的对比度以及与白色背景的对比度，缺点是无法输出至BBS代码。',
    contents: once(() => initializeColors(rawV2Colors)),
  },
  'bbs': {
    id: 'bbs',
    name: 'BBS颜色',
    description: '“BBS色板”包含18个常见于BBS的颜色，是传统的配色方案，缺点是色彩间的区分度及与背景的对比度无法保证。',
    contents: once(() => initializeColors(rawBbsColors)),
  },
}
