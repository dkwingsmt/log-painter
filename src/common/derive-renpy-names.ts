import forEach from 'lodash/forEach';
import size from 'lodash/size';
import max from 'lodash/max';
import isEmpty from 'lodash/isEmpty';

function within(code: number, lowestIncluded: number, highestIncluded: number): boolean {
  return code >= lowestIncluded && code <= highestIncluded;
}

function isValidStart(code: number): boolean {
  return within(code, 0x0041, 0x005A) // Uppercase letter ASCII
      || within(code, 0x0061, 0x007A) // Lowercase letter ASCII
      || within(code, 0x00C0, 0x00D6) // EASCII 1
      || within(code, 0x00D8, 0x00F6) // EASCII 2
      || within(code, 0x00F8, 0x02AF) // Before modifiers
      || within(code, 0x0370, 0x197F) // There are lots of exceptions but we ignore them for now
      || within(code, 0x1E00, 0x1FFF) // Latin extended, greek extended
      || within(code, 0x2160, 0x2188) // Roman numbers
      || within(code, 0x3040, 0x30FF) // Kanas
      || within(code, 0xAC00, 0xD7AF) // Hangul syllables
      // || code == 0x5F // _  // Python allows underscore but we reserve it for dedup.
      || within(code, 0x3400, 0x4DBF)   // CJK
      || within(code, 0x4e00, 0x9fff)   // CJK
      || within(code, 0xF900, 0xFAFF)   // CJK
      || within(code, 0x20000, 0x2CEAF) // CJK
      || within(code, 0x2F800, 0x2FA1F) // CJK
  ;
}

function isValidContinue(code: number): boolean {
  return isValidStart(code)
      || within(code, 0x0030, 0x0039)  // ASCII digits
      || within(code, 0x0300, 0x36F)   // Combining Diacritical Marks
      || code == 0x005F   // Low line: connector punc
      || code == 0x00B7   // Middle dot
  ;
}

const keywords: string[] = [
  'False',
  'await',
  'else',
  'import',
  'pass',
  'None',
  'break',
  'except',
  'in',
  'raise',
  'True',
  'class',
  'finally',
  'is',
  'return',
  'and',
  'continue',
  'for',
  'lambda',
  'try',
  'as',
  'def',
  'from',
  'nonlocal',
  'while',
  'assert',
  'del',
  'global',
  'not',
  'with',
  'async',
  'elif',
  'if',
  'or',
  'yield',
  'match',
  'case',
  '_',
];
const keywordsMap: Record<string, boolean> = {};
function isKeyword(s: string): boolean {
  if (isEmpty(keywordsMap)) {
    for (const keyword of keywords) {
      keywordsMap[keyword] = true;
    }
  }
  return keywordsMap[s] != null;
}

function sanitizeIdentifier(name: string): string {
  let metFirst = false;
  return name.split('').filter((c: string): boolean => {
    if (!metFirst) {
      return metFirst = isValidStart(c.codePointAt(0) ?? 0);
    } else {
      return isValidContinue(c.codePointAt(0) ?? 0);
    }
  }).join('');
}

export function deriveRenpyNames(names: string[]): string[] {
  console.log(JSON.stringify(names));
  let sourceDedupId = 0;
  const sourceDedupTable: Record<string, boolean> = {};
  // Map from "key" to "name"
  const sourceDeduped: string[] = [];
  for (const name of names) {
    if (sourceDedupTable[name] != null) {
      sourceDedupId += 1;
      sourceDeduped.push(`${name}${sourceDedupId}`);
    } else {
      sourceDeduped.push(name);
      sourceDedupTable[name] = true;
    }
  }
  console.log(JSON.stringify(sourceDeduped));

  // Map from "key" to "result"
  const finished: Record<string, string> = {};
  let dedupId = 0;
  function nextDedupName(): string {
    dedupId += 1;
    return `_C${dedupId}`;
  }

  // Map from "key" to "chars"
  let remaining: Record<string, string[]> = {};
  // Map from "filteredKey" to "key"
  const entireDedup: Record<string, string> = {};
  for (const name of sourceDeduped) {
    if (isEmpty(name)) {
      finished[name] = nextDedupName();
    } else {
      const sanitized = sanitizeIdentifier(name).split('');
      const existingKey = entireDedup[sanitized.join('')];
      if (sanitized.length == 0) {
        finished[name] = nextDedupName();
      } else if (existingKey != null) {
        finished[name] = nextDedupName();
        finished[existingKey] = nextDedupName();
        delete remaining[existingKey];
      } else {
        remaining[name] = sanitized;
        entireDedup[sanitized.join('')] = name;
      }
    }
  }
  const longest = max(Object.values(remaining).map((chars) => chars.length)) ?? 0;
  console.log('## Sanitized');
  console.log(remaining);

  // Map from "key" to "chars"
  let nextRemaining: Record<string, string[]> = {};

  for (let curLen = 1; curLen <= longest; curLen += 1) {
    if (isEmpty(remaining)) {
      break;
    }
    // console.log(`== Round ${curLen} ==`);
    // console.log(`Remaining: ${JSON.stringify(remaining)}`);

    // Map from "attempt" to "key"
    const working: Record<string, string> = {};
    // Test each entry in `remaining`. Record duplicates in `nextRemaining`.
    // Record non-dup results in `finished`.
    forEach(remaining, (chars: string[], key: string) => {
      // console.log(`key ${key} char ${chars}`);
      const attempt = chars.slice(0, curLen).join('');
      if (isKeyword(attempt)) {
        if (chars.length <= curLen) {
          finished[key] = nextDedupName();
        } else {
          nextRemaining[key] = chars;
        }
      } else if (working[attempt] == null) {
        working[attempt] = key;
        finished[key] = attempt; // Might be cleared later.
      } else {
        nextRemaining[key] = chars;
        nextRemaining[working[attempt]] = remaining[working[attempt]];
        delete finished[working[attempt]];
      }
    });
    remaining = nextRemaining;
    nextRemaining = {};
  }
  if (size(finished) != size(names)) {
    console.error('Algorithm error', names);
    throw Error('Algorithm error');
  }
  console.log(JSON.stringify(finished));
  return sourceDeduped.map((key) => finished[key]);
}
