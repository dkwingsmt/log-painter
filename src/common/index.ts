export { useStepperStyles } from './App-classes';

export { AlertDialog } from './AlertDialog';
export type { AlertDialogControl } from './AlertDialog';

export { loadRawConfig, saveRawConfig } from './storage';

export {
  MultiStep,
} from './MultiStep';
export type {
  StartStepProps,
  MiddleStepProps,
  EndStepProps,
} from './MultiStep';