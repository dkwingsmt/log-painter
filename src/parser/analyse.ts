import forEach from 'lodash/forEach';
import values from 'lodash/values';
import clone from 'lodash/clone';
import { ParseResult, ParsedPlayer } from './parser_chat';
import {
  Configuration,
  AnalysedLine,
  AnalysedPlayer,
  AnalyseResult,
  ConfigPlayer,
  presetDescribedColors,
  DescribedColor,
} from 'common';

interface GroupResult {
  players: AnalysedPlayer[];
  lines: AnalysedLine[];
};

const presetColors: string[] = presetDescribedColors.map(({ value }: DescribedColor) => value);

function getPlayerIdGroup(player: ParsedPlayer): string[] {
  const ids: string[] = [];
  if (player.number)
    ids.push(`qq:${player.number}`);
  ids.push(`name:${player.name}`);
  return ids;
}

function groupByPlayers(result: ParseResult): GroupResult {
  const lines: AnalysedLine[] = [];
  const players: Record<string, AnalysedPlayer> = {};
  for (const parsedLine of result.logLines) {
    const { player, time, content } = parsedLine;
    const playerIdGroup = getPlayerIdGroup(player);
    const majorPlayerId = playerIdGroup[0];
    if (!players[majorPlayerId])
      players[majorPlayerId] = {
        playerId: majorPlayerId,
        allPlayerIds: playerIdGroup,
        name: player.name,
        number: player.number,
      };
    lines.push({
      playerId: majorPlayerId,
      time,
      title: player.title,
      content,
    });
  }
  return {
    lines,
    players: values(players),
  };
}

function findFirstPlayerConfig(
  configPlayers: Record<string, ConfigPlayer>,
  playerIdGroup: string[],
): ConfigPlayer | null {
  for (const playerId of playerIdGroup) {
    if (configPlayers[playerId])
      return configPlayers[playerId];
  }
  return null;
}

function findFirstUnusedPresetColor(
  presetColors: string[],
  configPlayers: Record<string, ConfigPlayer>,
): string {
  const existingColors: Record<string, boolean> = {};
  forEach(configPlayers, (configPlayer: ConfigPlayer) => {
    existingColors[configPlayer.color] = true;
  });
  return presetColors.find((color: string) => !existingColors[color]) || 'black';
}

export function analyse(parsed: ParseResult, currentConfig: Configuration): AnalyseResult {
  const { lines, players } = groupByPlayers(parsed);
  const configPlayers: Record<string, ConfigPlayer> = clone(currentConfig.players);
  for (const player of players) {
    let configPlayer = findFirstPlayerConfig(configPlayers, player.allPlayerIds);
    if (!configPlayer) {
      configPlayer = {
        enabled: true,
        displayName: player.name,
        color: findFirstUnusedPresetColor(presetColors, configPlayers),
      };
    }
    for (const playerId of player.allPlayerIds) {
      configPlayers[playerId] = configPlayer;
    }
  };
  return {
    lines,
    playerIds: players.map((player: AnalysedPlayer) => player.playerId),
    nextConfig: {
      players: configPlayers,
    },
  };
}