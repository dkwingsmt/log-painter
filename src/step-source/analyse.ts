import {
  AnalysedPlayer,
  GroupResult,
  AnalysedLine,
} from 'step-source';
// import {
//   bbsColors,
//   DescribedColor,
// } from './colors';
import {
  PlayerConfig,
} from 'common';

export interface AnalyseResult {
  lines: AnalysedLine[];
  playerIds: string[];
}

// const presetColors: string[] = bbsColors.map(({ value }: DescribedColor) => value);

function findFirstPlayerConfig(
  configPlayers: Record<string, PlayerConfig>,
  playerIdGroup: string[],
): PlayerConfig | null {
  for (const playerId of playerIdGroup) {
    if (configPlayers[playerId])
      return configPlayers[playerId];
  }
  return null;
}

export function analyse(grouped: GroupResult): AnalyseResult {
  const { players, lines } = grouped;
  const configPlayers: Record<string, PlayerConfig> = {};
  const currentPlayers: Record<string, PlayerConfig> = {};
  for (const player of players) {
    let configPlayer = findFirstPlayerConfig(configPlayers, player.allPlayerIds);
    if (!configPlayer) {
      configPlayer = {
        enabled: true,
        displayName: player.name,
        color: 'black',
      };
    }
    for (const playerId of player.allPlayerIds) {
      currentPlayers[playerId] = configPlayer;
    }
  };
  return {
    lines,
    playerIds: players.map((player: AnalysedPlayer) => player.playerId),
  };
}