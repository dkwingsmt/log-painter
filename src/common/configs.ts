import React from 'react';
import Store from 'store';

import { ColorPalette, colorPalettes } from './colors';

export interface GeneralConfig {
  removeLinesStartedWithParenthesis: boolean;
  removeLinesStartedWithDot: boolean;
  removeLinesStartedWithLenticular: boolean;
  regularizeQuotes: boolean;
  palette: ColorPalette;
}

export const defaultGeneralConfig: GeneralConfig = {
  removeLinesStartedWithParenthesis: false,
  removeLinesStartedWithDot: false,
  removeLinesStartedWithLenticular: false,
  regularizeQuotes: false,
  palette: 'v2',
};

export interface PlayerConfig {
  displayName: string;
  enabled: boolean;
  color: string;
}

export interface Configuration {
  players: Record<string, PlayerConfig>;
  general: GeneralConfig;
}

export const sanitizeConfig = (value: {} | undefined): Configuration => {
  const nonNullValue = (value ?? {}) as Record<string, {}>;
  const result = {
    players: (nonNullValue['players'] ?? {}) as Record<string, PlayerConfig>,
    general: {
      ...defaultGeneralConfig,
      ...nonNullValue['general'] as GeneralConfig
    },
  };
  if (!(result.general.palette in colorPalettes)) {
    result.general.palette = defaultGeneralConfig.palette;
  }
  return result;
};

const emptyConfig = sanitizeConfig(undefined);

export interface ConfigStorage {
  load: () => Configuration | undefined;
  save: (value: Configuration) => void;
}

export const realConfigStorage: ConfigStorage = {
  load: () => {
    return Store.get('config');
  },
  save: (config: {}): void => {
    Store.set('config', config);
  },
};

export const configContext: React.Context<Configuration> = React.createContext<Configuration>(emptyConfig);
