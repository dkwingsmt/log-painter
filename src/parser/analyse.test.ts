import { analyse } from './analyse';
import { Configuration, GroupResult, AnalysedPlayer, AnalyseResult } from 'common';

it('should work for empty config', () => {
  const player1: AnalysedPlayer = {
    playerId: 'name:a',
    allPlayerIds: ['name:a'],
    name: 'a',
  };
  const player2: AnalysedPlayer = {
    playerId: 'name:b',
    allPlayerIds: ['name:b'],
    name: 'b',
  };
  const groupResult: GroupResult = {
    players: [
      player1,
      player2,
    ],
    lines: [
      {
        time: 'time1',
        playerId: player1.playerId,
        content: ['1', '2'],
        title: 'A',
      },
      {
        time: 'time2',
        playerId: player2.playerId,
        content: ['3'],
        title: 'B',
      },
      {
        time: 'time3',
        playerId: player1.playerId,
        content: ['4'],
        title: 'A',
      },
    ]
  };
  const oldConfig: Configuration = {
    players: {},
  };
  const analyseResult: AnalyseResult = analyse(groupResult, oldConfig);
  expect(analyseResult).toMatchSnapshot();
});

it('should work when some players have filled starting colors', () => {
  const player1: AnalysedPlayer = {
    playerId: 'name:a',
    allPlayerIds: ['name:a'],
    name: 'a',
  };
  const player2: AnalysedPlayer = {
    playerId: 'name:b',
    allPlayerIds: ['name:b'],
    name: 'b',
  };
  const groupResult: GroupResult = {
    players: [
      player1,
      player2,
    ],
    lines: [
      {
        time: 'time1',
        playerId: player1.playerId,
        content: ['1', '2'],
        title: 'A',
      },
      {
        time: 'time2',
        playerId: player2.playerId,
        content: ['3'],
        title: 'B',
      },
    ]
  };
  const oldConfig: Configuration = {
    players: {
      'name:a': {
        enabled: true,
        displayName: 'A',
        color: 'black',
      },
      'name:d': {
        enabled: true,
        displayName: 'D',
        color: 'silver',
      },
    },
  };
  const analyseResult: AnalyseResult = analyse(groupResult, oldConfig);
  expect(analyseResult).toMatchSnapshot();
});