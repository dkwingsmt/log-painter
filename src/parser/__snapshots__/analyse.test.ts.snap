// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should work for empty config 1`] = `
Object {
  "lines": Array [
    Object {
      "content": Array [
        "1",
        "2",
      ],
      "playerId": "name:a",
      "time": "time1",
      "title": "A",
    },
    Object {
      "content": Array [
        "3",
      ],
      "playerId": "name:b",
      "time": "time2",
      "title": "B",
    },
    Object {
      "content": Array [
        "4",
      ],
      "playerId": "name:a",
      "time": "time3",
      "title": "A",
    },
  ],
  "nextConfig": Object {
    "players": Object {
      "name:a": Object {
        "color": "black",
        "displayName": "a",
        "enabled": true,
        "playerId": "name:a",
      },
      "name:b": Object {
        "color": "silver",
        "displayName": "b",
        "enabled": true,
        "playerId": "name:b",
      },
    },
  },
  "playerIds": Array [
    "name:a",
    "name:b",
  ],
}
`;
